File: Leetcode/Python Solutions/Stack/ValidParentheses.py
"""
LeetCode Problem: 20. Valid Parentheses
Link: https://leetcode.com/problems/valid-parentheses/
Language: Python
Written by: Mostofa Adib Shakib

Time Compplexity: O(n)
Space Complexity: O(n)
"""

class Solution:
    def isValid(self, s: str) -> bool:
        hashmap = {'(': ')', '{':'}', '[':']'}
        stack = []
        
        for i in s:
            if i in hashmap:
                stack.append(i)
                
            elif len(stack) > 0 and i == hashmap[stack[-1]]:
                stack.pop()
                
            else:
                return False
                
        return stack == []

File: Leetcode/Java Solutions/ValidParentheses.java
/*
LeetCode Problem: 20. Valid Parentheses
Link: https://leetcode.com/problems/valid-parentheses/
Written by: Mostofa Adib Shakib
Language: Java
*/

class Solution {
    public boolean isValid(String s) {
        HashMap<Character, Character> hashmap = new HashMap<Character, Character>();
        hashmap.put('(',')');
        hashmap.put('[', ']');
        hashmap.put('{', '}');
        
        Stack<Character> stack = new Stack<Character>();   
        int length = s.length();
        
        for (int i = 0; i < length; i++) {
            
            if ( s.charAt(i) == '(' || s.charAt(i) == '[' || s.charAt(i) == '{' ) {
                stack.push(s.charAt(i));
            }
            
            else {
                
                if ( stack.size() > 0 && hashmap.get(stack.peek()) == s.charAt(i) ) {
                    stack.pop();
                }
                else {
                    return false;
                }
            }
        }
        return stack.size() == 0 ? true:false; 
    }
}